### ISSUE TITLE
Custom multiscale level

### ISSUE THREAD

**Issue by ClementCaporal** [Napari/napari#6418]
## 🚀 Feature
Let the user choose the lowest resolution in multiscale image depending of the requested_shape.

## Motivation
From https://github.com/napari/napari/blob/47b9527ef59bf1b611a6b7f9e3c7877492af3f4a/napari/layers/utils/layer_utils.py#L561C1-L566C23
```python
   """ The level of the multiscale should be the lowest resolution such that
    the requested shape is above the shape threshold. By passing a shape
    threshold corresponding to the shape of the canvas on the screen this
    ensures that we have at least one data pixel per screen pixel, but no
    more than we need."""
```
I think this is perfect if the data is fast enough to load. When it is slow, I prefer to have control to force a lowest resolution to be loaded "longer" to explore my data quicker even if a better resolution could fit the screen.

## Pitch

I don't know the best way to do it.
My naive idea would be to 
- add a parameter to [`compute_multiscale_level`](
https://github.com/napari/napari/blob/47b9527ef59bf1b611a6b7f9e3c7877492af3f4a/napari/layers/utils/layer_utils.py#L522). For example `resolution_downscale` that will modify `scale_shape`
- add an attribute to the canvas `resolution_downscale` that can be adjusted be the user

## Alternatives

- Hack the `downsample_factors` when loading the multiscale data but it will make the scale bar wrong
- Use a screen with large physical pixels?

## Additional context

Quickly discussed with @melonora

**Comment by jni** [comment]
I think this becomes unnecessary once #5561 gets merged — it *should* just load low res first and then progressively load higher resolutions. Right @kephale? 😅

**Comment by kephale** [comment]
Probably, but... one thing I'm seeing with some frequency in the data/workflows that I'm looking at now (in cryoET land) is different algorithms run on different scale levels. I'm trying to understand if those cases may require the user to set the target resolution level. The technical implementation in https://github.com/napari/napari/issues/5561 can be relatively easily adapted, but I also don't want to overengineer it. In my case, I'm running segmentation algorithms that operate on low res data, then the results are used for processing the high resolution data, so there is a need to inspect both the low and high res data.

@aganders3, we had talked about working on squishing the scale levels in the progressive loading implementation into a single layer. This makes me wonder if we might want to actually expose some controls about scale levels on those squished layers.

**Comment by psobolewskiPhD** [comment]
> In my case, I'm running segmentation algorithms that operate on low res data, then the results are used for processing the high resolution data, so there is a need to inspect both the low and high res data.

This is pretty common in the QuPath whole-slide world. Pixel classification and cell segmentation can be performed at different scale levels, with some tradeoffs of speed and precision of course.

**Comment by aganders3** [comment]
Likewise I have made use of [multiscale edge detection](https://www.sciencedirect.com/science/article/abs/pii/S0950705113000464) before, and it would make sense take advantage of precomputed multiscale data where possible.

This feels like two separate things though: processing and viewing. I can see a benefit in giving better control or access for each. For viewing I agree the progressive loading will get a lot of the way there, but I can imagine a preference to load all chunks before displaying any higher res. Even still users may want to limit to a certain resolution max for performance reasons, depending on their hardware or how the data is chunked for example.

**Comment by kephale** [comment]
Yes.

I'm bringing up the processing aspect with respect to viewing because of the need to inspect results in conjunction with the scale levels that were used to generate them. If I'm missing a particle in my segmentation that was run on a low-res scale, then I want to visualize/know if it is because the downscaling averaged away the signal or if the segmentation algorithm is failing.

It does sound like there are a handful of reasons to have more control over how scales are loaded.

**Comment by psobolewskiPhD** [comment]
Bounced back here but a layer group for multiscale would be nice, where you get the normal function where scales are auto-swapped with the group closed, but can "open"/"expand it" and access each of the scales. Could be very useful again in cases where you want to do training or segmentation on less data.

**Comment by psobolewskiPhD** [comment]
Ran into this again, so bumping. Most plugins use Layer.data and don't handle multiscale data properly.
It's just assumed that Layer.data is an array or array-like, but for multiscale layers it's a list.
Probably need some better documentation to plugin authors to raise awareness, but also probably need some UI options to make it easier to expose multiscale level, so the user can run something like a skimage threshold on a specific level. This would need to be coupled with a smooth way to ensure the resultant derived layer is mapped into the multiscale world properly.
Alternately, we'd need some sort of mechanism to process the levels of a multiscale layer as arrays, in parallel or series.

**Comment by cmalinmayor** [comment]
I just found this issue when hoping to let the user manually set the resolution level in 3D. I have a minimal working example [on a fork](https://github.com/napari/napari/compare/main...cmalinmayor:napari:multiscale-level-lock) that adds a dropdown to layer controls for multiscale image/segmentation layers that lets the user pick the resolution level, or go back to auto-select, in 2D and 3D.

<img width="960" height="618" alt="Image" src="https://github.com/user-attachments/assets/3876f653-92a4-4423-9dd7-e4f739e651e9" />

Needs cleanup and testing, but let me know if there is interest in merging this or not. I know we are working on a more complex 3D approach, but at least as a stopgap that allows user control, this would be nice. I'm even fine with no UI component as long as there is an API for doing it - currently the hard coded selection of the lowest resolution in multiple places with -1 indexing prevents any attempt to work around this programatically.

@kephale and I discussed this briefly - anything to add?

```
# /// script
# requires-python = ">=3.10"
# dependencies = [
#     "napari[all] @ git+https://github.com/cmalinmayor/napari.git@multiscale-level-lock",
#     "numpy",
# ]
# ///
"""Test script for the locked_data_level feature with the Qt widget.

Creates a 3D multiscale image with 4 resolution levels and opens it in napari.
Each level has a distinct value so you can verify which level is being rendered.
"""

import numpy as np
import napari

# Build a 4-level multiscale pyramid (3D)
shapes = [(128, 256, 256), (64, 128, 128), (32, 64, 64), (16, 32, 32)]
multiscale_data = []
for i, shape in enumerate(shapes):
    # Fill each level with a distinct intensity so it's obvious which is loaded
    arr = np.random.randint(0, 80, size=shape, dtype=np.uint8)
    # Add a bright marker whose size scales with resolution so you can tell levels apart
    s = shape[0] // 4
    arr[s : s * 2, s : s * 2, s : s * 2] = 50 * (i + 1)
    multiscale_data.append(arr)

viewer = napari.Viewer(ndisplay=3)
layer = viewer.add_image(multiscale_data, name="test_multiscale", multiscale=True)

print("Multiscale layer added.")
print(f"  Levels: {[a.shape for a in multiscale_data]}")
print()
print("Use the 'data level' dropdown in the layer controls to lock a level.")
print("  'Auto' = default napari behavior (lowest res in 3D)")
print("  '0: ...' = full resolution")
print("  '3: ...' = coarsest resolution")
print()
print("You can also set it programmatically:")
print("  viewer.layers['test_multiscale'].locked_data_level = 0")
print("  viewer.layers['test_multiscale'].locked_data_level = None  # restore auto")

napari.run()
```

**Comment by psobolewskiPhD** [comment]
@cmalinmayor Great timing, I was also thinking about something like this that allows a user to specific a level explicitly, even for 2D it can be useful e.g. to avoid loading if only a specific res level is needed (e.g. proof reading a segmentation).
Makes even more sense for 3D. 
I've not looked at your code, but to my eye your video is exactly what I was thinking, so I'd say open a PR and we go from there?

**Comment by TimMonko** [comment]
@cmalinmayor this is incredible!
I wouldn't even think of it as a stopgap, this is a real nice way to expose multi-levels. In fact, I think it will make explaining how multi-level works MUCH easier since it's hard to intuit the zoom-level-changes-multiscale-level when I'm teaching napari. Being able to just select like a low res data level while being zoomed in sounds perfect and useful, and for that reason I'd want to keep this around forever, especially since it slots in so simply to the layer controls :)

**Comment by psobolewskiPhD** [comment]
I wonder how we can get `MultiScaleData.__array__` to use the "forced" data_level instead of the current -1.
Would be very cool if the plugins that don't know about MultiScaleData would get what the user selected as the level when they use Layer.data.

**Comment by cmalinmayor** [comment]
> In fact, I think it will make explaining how multi-level works MUCH easier since it's hard to intuit the zoom-level-changes-multiscale-level when I'm teaching napari. 

@TimMonko I think you'll like the new demo I made in the PR even more for teaching purposes 😃

**Comment by cmalinmayor** [comment]
> I wonder how we can get `MultiScaleData.__array__` to use the "forced" data_level instead of the current -1. Would be very cool if the plugins that don't know about MultiScaleData would get what the user selected as the level when they use Layer.data.

@psobolewskiPhD I did slightly look into this, but since `MultiScaleData` doesn't know anything about the Layer's `data_level` setting, it doesn't fit naturally into the `__array__` API. `Layer.data` could select the level based on the layer's selected data level, but it would be a highly breaking change for anyone who expects a MultiScaleData object

**Comment by psobolewskiPhD** [comment]
Darn, I figured that would probably the case, but I was secretly hoping I was wrong and someone would spot a clever way to handle it. Maybe we can make a follow-up issue so we don't lose this thought once this closes out.

Linking this PR as well:
https://github.com/napari/napari/pull/8822

Feedback thoughts etc. on MultiScaleData (either here or in that PR) from you and @kephale would be very welcome.
